### PR TITLE
Mv bucket expiry2

### PR DESCRIPTION
**Pull request by matthewvon**

### CODE REVIEW

File: src/eleveldb_sup.erl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+%% -------------------------------------------------------------------
+-module(eleveldb_sup).
+
+-behaviour(supervisor).
+
+-export([start_link/0]).
+-export([init/1]).
+
+start_link() ->
+    supervisor:start_link({local, ?MODULE}, eleveldb_sup, []).
```
Comment (tburghart) [inline src/eleveldb_sup.erl]: This name should be added to the `registered` list in the .app.src file.

File: src/eleveldb_app.erl [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+-behavior(application).
+
+-export([start/2, stop/1]).
+
+start(_Type, _Args) ->
+    case eleveldb_sup:start_link() of
+        {ok, Pid} ->
+            {ok, Pid};
+        Other ->
```
Comment (tburghart) [inline src/eleveldb_app.erl]: Either return the result of `eleveldb_sup:start_link` directly, which is standard practice, or handle the possible returns from `supervisor:start_link`.
As written, `{error, {error, ...}}` is a possible result, which is undesirable for upstream reporting.

File: src/eleveldb.erl [hunk 3]
```diff
@@ -335,7 +338,13 @@ validate_options(Type, Opts) ->
                             validate_type(KType, V)
                     end, Opts).
 
+-spec property_cache(string(), string()) -> ok.
+property_cache(_BucketKey, _Properties) ->
+    erlang:nif_error({error, not_loaded}).
 
+-spec set_metadata_pid(pid()) -> ok.
+set_metadata_pid(_Pid) ->
+    erlang:nif_error({error, not_loaded}).
```
Comment (tburghart) [inline src/eleveldb.erl]: I would prefer a generic interface here to register an external service pid so we don't need a new function each time we do this sort of thing, and I don't expect this to be the only service we're going to want to call from eleveldb (think lager).

For example:

```erlang
-spec register_service(Type :: atom(), Pid :: pid(), Info :: term()) -> {ok, reference()} | {error, term()}.
-spec unregister_service(Type :: atom(), Key :: pid() | reference()) -> ok | {error, term()}.
```

In this example,
```erlang
    set_metadata_pid(P)
```
might be called as
```erlang
    register_service(eleveldb_metadata, P, [])
```

I'm not sure about the `unregister` side, just throwing something out for the case where we detect some service is unavailable. It seems a ref could be associated with a registration that could subsequently be used to unregister that service.

File: src/eleveldb.app.src [hunk 4]
```diff
@@ -22,5 +22,6 @@
          %%  "older" compression by default
          {compression, snappy}
 
-        ]}
+        ]},
+   {mod, {eleveldb_app, []}}
  ]}.
```
Comment (tburghart) [inline src/eleveldb.app.src]: Add named servers started by the application to the `registered` list.

File: c_src/eleveldb.cc [hunk 5]
```diff
@@ -73,13 +78,17 @@ static ErlNifFunc nif_funcs[] =
     {"async_iterator", 3, eleveldb::async_iterator},
     {"async_iterator", 4, eleveldb::async_iterator},
 
-    {"async_iterator_move", 3, eleveldb::async_iterator_move}
+    {"async_iterator_move", 3, eleveldb::async_iterator_move},
+
+    {"property_cache", 2, eleveldb::property_cache},
+    {"set_metadata_pid", 1, eleveldb::set_metadata_pid}
```
Comment (tburghart) [inline c_src/eleveldb.cc]: Same as my comment in erleveldb.erl, I'd like this to be a more generalized mechanism.

File: src/eleveldb_metadata.erl [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            %% No recursion here; `system_continue/3' will be invoked
+            %% if appropriate
+            sys:handle_system_msg(Request, From, Parent,
+                                  ?MODULE, Debug, Fun);
+
+        callback_shutdown=Msg ->
+            sys:handle_debug(Debug, fun write_debug/3,
+                             ?MODULE,
+                             {in, Msg}),
+            ok;
```
Comment (tburghart) [inline src/eleveldb_metadata.erl]: Should this not call `erlang:exit(shutdown)`?

File: src/eleveldb_metadata.erl [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+-module(eleveldb_metadata).
+
```
Comment (tburghart) [inline src/eleveldb_metadata.erl:23]: I realize this is designed to handle raw messages from a NIF, but a protocol interface with lots and lots of doc comments seems like a good thing to have.

File: src/eleveldb_sup.erl [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+init(_Args) ->
+    SupFlags = {one_for_one, 2, 5},
+    ChildSpecs = [{metadata, {eleveldb_metadata, start_link, []},
+                   permanent, brutal_kill, worker, [eleveldb_metadata]}],
```
Comment (tburghart) [inline src/eleveldb_sup.erl]: Why `brutal_kill`?  The process should be written to handle exit messages cleanly.

File: src/eleveldb_metadata.erl [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+-export([callback_router/3]).
+
+%% Application messages are sent to this process directly from
+%% NIF-land instead of other Erlang modules, so this does not
+%% implement an OTP behavior.
+%%
+%% See http://erlang.org/doc/design_principles/spec_proc.html for
+%% information on these system message handlers.
+-export([system_continue/3, system_terminate/4,
```
Comment (tburghart) [inline src/eleveldb_metadata.erl]: If this were a `gen_whatever` shutdown messages could unregister the service, as I've suggested elsewhere.

**Review by tburghart — DISMISSED** [review]
I haven't gone into the C++ side at all, as I thing significant changes are needed on the Erlang side first.

So, this creates a circular dependency on `riak_core` (see xref output), which is a Very Bad Thing™, suggesting (strongly) that the model is wrong and that the metadata service should be running in `riak_core`, not in `eleveldb`.
Obviously, `eleveldb` needs to support registration of the service, but it doesn't need to run the service itself.

Needs unit tests, LOTS of them!  (https://github.com/basho/eleveldb/pull/240 might make that easier, were it to be merged)

At a minimum, and I'm not saying this is the right thing to do, a clean build requires:
* Dialyzer apps list `Makefile`.
* XRef exclusions in `rebar.config`.

**Comment by tburghart** [comment]
@JeetKunDoug is there an associated PR in `riak_core`?  I see what I assume is your working branch, a review-only PR might be good.

**Comment by JeetKunDoug** [comment]
Yea - was waiting to hear from @matthewvon about the viability of it before I PRed the riak_core change, but can WIP PR for review now...

**Comment by JeetKunDoug** [comment]
riak_core changes are in basho/riak_core#898

**Comment by tburghart** [comment]
Grrr, GitHub makes it look so simple ... do _NOT_ let it resolve the conflicts for you!

The `mv-bucket-expiry2-try-merge` branch has what you get if you do, and appears to be pulling in more than expected.

**Comment by macintux** [comment]
There will need to be changes to reflect the `riak_core` code that's likely to be merged soon.

At the very least, the callback functions that handle the PID registration and receive the bucket properties must now return `ok`.

https://github.com/basho/riak_core/pull/898 for more details, especially the commentary at the top of https://github.com/basho/riak_core/blob/dr-th/service-poc/src/riak_core_info_service.erl

**Comment by macintux** [comment]
* https://github.com/basho/riak_core/pull/898 <- merged
* https://github.com/basho/riak_core/pull/904 <- pending, adds `bucket_props` argument to the pid-handling callbacks

**Comment by erikleitch** [comment]
(I thought @tburghart was reviewing this one, but the riak team said this morning that I should look at it too, so I'm adding a few notes here, despite not being sure whether this is ready for review, etc)

---
* **eleveldb_metadata.erl** Is there a deep reason why
    eleveldb:property_cache() is separately called from
    eleveldb_metadata.erl (instead of just having riak_core call
    `eleveldb:property_cache` directly)?  (Or is it leftover from an
    earlier architecture that has now changed?)

* In **router.cc**, is there a reason not to return an error atom (or
    message) if any of `add_metadata_pid`, `remove_metadata_pid`, or
    `property_cache` is called incorrectly? (currently, the code
    returns ATOM_OK for a number of these, whether or not there's an
    error).

  (Looks like these are in process of changing, so maybe this comment
  will become irrelevant)

* **router.cc** uses a global service callback object
  (`gBucketPropCallback`) to maintain the erlang pid to contact for
  bucket props.  This is used both from leveldb to request bucket prop
  info (via `leveldb_callback`) and from erlang to add/remove the pid
  (via `set_metadata_pid/remove_metadata_pid`), so simultaneous access
  from multiple thread contexts is possible.

  Shouldn't this resource be mutex-protected to ensure that the pid
  isn't accessed after `remove_metadata_pid` has been called, or that
  `ServiceCallback::GetPid` returns true after the pid has been set?

**Comment by macintux** [comment]
Good questions Erik.

* I think it makes sense to have a translation function between the riak_core callback data and what the NIF expects. The former was designed to be general-purpose.
* `eleveldb_metadata` in general is a relic of an earlier architecture, so we could move that translation function into `eleveldb` proper, although for the moment I'd like to leave it so we don't have to yet again touch the riak_core PR.
* The NIFs need to return `ok` if, and only if, riak_core should continue servicing eleveldb metadata queries. The riak_core info service code shuts down the service process if any other return value is received. So, if there is an error that means that leveldb can't use the info service, then return an error.

**Comment by macintux** [comment]
Ideally I think we'd allow eleveldb to report two categories of errors: permanent and transient. Future enhancement to the info service, perhaps.

**Comment by matthewvon** [comment]
Added locking around ServiceCallback accessors per code review.

**Review by erikleitch — APPROVED** [review]

**Review by erikleitch — COMMENTED** [review]
Oops - I think the spec for `eleveldb:property_cache/2` is now wrong (`eleveldb::property_cache` changed to return ok, or {error, any()}, but spec still reads: `-spec property_cache(string(), string()) -> ok.`)

**Comment by matthewvon** [comment]
Wrong, maybe. In flux, absolutely.  Waiting on API lock down.

**Comment by matthewvon** [comment]
This was a horrible merge.  Done by hand directly to develop.

**Pull request closed without merging**